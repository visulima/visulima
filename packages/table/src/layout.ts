import type { Cell as CellType, LayoutCell, TableLayout } from "./types";

/** Creates a layout cell from a given cell configuration. */
const createLayoutCell = (cell: CellType, column: number, row: number): LayoutCell => {
    const normalizedCell = typeof cell === "object" && cell !== null ? cell : { content: String(cell) };

    return {
        ...normalizedCell,
        content: String(normalizedCell.content ?? ""),
        height: normalizedCell.rowSpan ?? 1,
        width: normalizedCell.colSpan ?? 1,
        x: column,
        y: row,
    };
}

/**
 * Creates a list of layout cells from a 2D array of rows (CellType[][]).
 * Also inserts "span cells" (placeholders) for the covered columns/rows.
 */
// eslint-disable-next-line sonarjs/cognitive-complexity
const createTableLayout = (rows: CellType[][]): TableLayout => {
    // Step 1: figure out the total columns by max sum of colSpans in any row
    let maxCols = 0;
    for (const row of rows) {
        let sum = 0;
        for (const cell of row) {
            if (cell === null || cell === undefined) {
                sum += 1;
            } else if (typeof cell === "object" && !Array.isArray(cell)) {
                sum += cell.colSpan ?? 1;
            } else {
                sum += 1;
            }
        }
        maxCols = Math.max(maxCols, sum);
    }

    // Step 2: build out the cells
    const layoutCells: LayoutCell[] = [];
    let rowIndex = 0;
    
    // Create a grid to track occupied positions
    const occupiedPositions: Record<string, LayoutCell> = {};
    
    // Helper function to check if a position is occupied
    const isPositionOccupied = (x: number, y: number): boolean => {
        return occupiedPositions[`${x},${y}`] !== undefined;
    };
    
    // Helper function to mark a position as occupied
    const markPositionOccupied = (x: number, y: number, cell: LayoutCell): void => {
        occupiedPositions[`${x},${y}`] = cell;
    };

    for (const row of rows) {
        // We track where we place each cell. colPointer moves left to right.
        let colPointer = 0;
        for (const cellValue of row) {
            if (cellValue == null) {
                // This 'slot' is presumably covered by a spanning cell
                colPointer += 1;
                continue;
            }
            
            // Skip over any positions already occupied by spans from earlier rows
            while (colPointer < maxCols && isPositionOccupied(colPointer, rowIndex)) {
                colPointer += 1;
            }
            
            // If we reached the end of the row, break out
            if (colPointer >= maxCols) {
                break;
            }
            
            // Create a layout cell for this position
            const layoutCell = createLayoutCell(cellValue, colPointer, rowIndex);
            layoutCells.push(layoutCell);
            
            // Mark all positions covered by this cell as occupied
            for (let ry = rowIndex; ry < rowIndex + layoutCell.height; ry++) {
                for (let rx = colPointer; rx < colPointer + layoutCell.width; rx++) {
                    markPositionOccupied(rx, ry, layoutCell);
                }
            }

            // Insert placeholder cells for all covered positions except the top-left
            for (let ry = rowIndex; ry < rowIndex + layoutCell.height; ry++) {
                for (let rx = colPointer; rx < colPointer + layoutCell.width; rx++) {
                    if (rx === colPointer && ry === rowIndex) {
                        // the real cell - already added
                        continue;
                    }
                    // placeholder cell
                    layoutCells.push({
                        content: "",
                        height: 1,
                        isSpanCell: true,
                        parentCell: layoutCell,
                        width: 1,
                        x: rx,
                        y: ry,
                    });
                }
            }

            colPointer += layoutCell.width;
        }
        rowIndex++;
    }

    // Step 3: total table height is the number of rows
    // (But some cells might extend downward. We'll find the max.)
    let maxRow = rows.length;
    for (const c of layoutCells) {
        maxRow = Math.max(maxRow, c.y + c.height);
    }

    // Step 4: return the structure
    return {
        cells: layoutCells,
        height: maxRow,
        width: maxCols,
    };
}

export default createTableLayout;
