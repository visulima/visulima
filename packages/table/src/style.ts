import type { BorderStyle } from "./types";

/**
 * Default border style using simple ASCII characters.
 */
export const DEFAULT_BORDER: BorderStyle = {
    bodyJoin: "│",
    bodyLeft: "│",
    bodyRight: "│",
    bottomBody: "─",
    bottomJoin: "┴",
    bottomLeft: "└",
    bottomRight: "┘",
    joinBody: "─",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "─",
    topJoin: "┬",
    topLeft: "┌",
    topRight: "┐",
};

/**
 * Minimal border style using only horizontal lines.
 */
export const MINIMAL_BORDER: BorderStyle = {
    bodyJoin: "",
    bodyLeft: "",
    bodyRight: "",
    bottomBody: "-",
    bottomJoin: "-",
    bottomLeft: "-",
    bottomRight: "-",
    joinBody: "-",
    joinJoin: "-",
    joinLeft: "-",
    joinRight: "-",
    topBody: "-",
    topJoin: "-",
    topLeft: "-",
    topRight: "-",
};

/**
 * Double-line border style using Unicode box-drawing characters.
 */
export const DOUBLE_BORDER: BorderStyle = {
    bodyJoin: "║",
    bodyLeft: "║",
    bodyRight: "║",
    bottomBody: "═",
    bottomJoin: "╩",
    bottomLeft: "╚",
    bottomRight: "╝",
    joinBody: "═",
    joinJoin: "╬",
    joinLeft: "╠",
    joinRight: "╣",
    topBody: "═",
    topJoin: "╦",
    topLeft: "╔",
    topRight: "╗",
};

/**
 * Border style with rounded corners using Unicode box-drawing characters.
 */
export const ROUNDED_BORDER: BorderStyle = {
    bodyJoin: "│",
    bodyLeft: "│",
    bodyRight: "│",
    bottomBody: "─",
    bottomJoin: "┴",
    bottomLeft: "╰",
    bottomRight: "╯",
    joinBody: "─",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "─",
    topJoin: "┬",
    topLeft: "╭",
    topRight: "╮",
};

/**
 * Border style using dots for the border.
 */
export const DOTS_BORDER: BorderStyle = {
    bodyJoin: "┊",
    bodyLeft: "┊",
    bodyRight: "┊",
    bottomBody: "┈",
    bottomJoin: "┴",
    bottomLeft: "└",
    bottomRight: "┘",
    joinBody: "┈",
    joinJoin: "┼",
    joinLeft: "├",
    joinRight: "┤",
    topBody: "┈",
    topJoin: "┬",
    topLeft: "┌",
    topRight: "┐",
};

/**
 * Border style using Markdown syntax.
 */
export const MARKDOWN_BORDER: BorderStyle = {
    bodyJoin: "|",
    bodyLeft: "|",
    bodyRight: "|",
    bottomBody: "-",
    bottomJoin: "|",
    bottomLeft: "|",
    bottomRight: "|",
    joinBody: "-",
    joinJoin: "|",
    joinLeft: "|",
    joinRight: "|",
    topBody: "-",
    topJoin: "|",
    topLeft: "|",
    topRight: "|",
};

/**
 * ASCII border style using only ASCII characters.
 */
export const ASCII_BORDER: BorderStyle = {
    bodyJoin: "|",
    bodyLeft: "|",
    bodyRight: "|",
    bottomBody: "-",
    bottomJoin: "+",
    bottomLeft: "+",
    bottomRight: "+",
    joinBody: "-",
    joinJoin: "+",
    joinLeft: "+",
    joinRight: "+",
    topBody: "-",
    topJoin: "+",
    topLeft: "+",
    topRight: "+",
};

/**
 * No border style.
 */
export const NO_BORDER: BorderStyle = {
    bodyJoin: "",
    bodyLeft: "",
    bodyRight: "",
    bottomBody: "",
    bottomJoin: "",
    bottomLeft: "",
    bottomRight: "",
    joinBody: "",
    joinJoin: "",
    joinLeft: "",
    joinRight: "",
    topBody: "",
    topJoin: "",
    topLeft: "",
    topRight: "",
};
