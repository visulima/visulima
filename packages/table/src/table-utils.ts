import type { Cell as CellType , LayoutCell } from "./types";

/** Computes the logical width of a row (the sum of colSpans, defaulting to 1 per cell). */
export function computeRowLogicalWidth(row: CellType[]): number {
    // eslint-disable-next-line unicorn/no-array-reduce
    return row.reduce((total: number, cell: CellType) => {
        if (cell === null) {
            return total + 1;
        }

        if (typeof cell === "object" && !Array.isArray(cell)) {
            return total + (cell.colSpan ?? 1);
        }

        return total + 1;
    }, 0);
}

/** Pads a row with null cells so that its logical width equals targetWidth. */
export function fillRowToWidth(row: CellType[], targetWidth: number): CellType[] {
    const currentWidth = computeRowLogicalWidth(row);

    if (currentWidth < targetWidth) {
        return [...row, ...Array.from({length: targetWidth - currentWidth}).fill(null)] as CellType[];
    }

    return row;
}

/** Returns the underlying "real" cell (if a span cell, returns its parent). */
export function getRealCell(layoutCell: LayoutCell | null): LayoutCell | null {
    if (!layoutCell) {
        return null;
    }

    if (layoutCell.isSpanCell && layoutCell.parentCell) {
        return layoutCell.parentCell;
    }

    return layoutCell;
}

/** Checks whether two layout cells represent the same originating cell. */
export function areCellsEquivalent(cellA: LayoutCell | null, cellB: LayoutCell | null): boolean {
    const realA = getRealCell(cellA);
    const realB = getRealCell(cellB);
    return realA === realB;
}
