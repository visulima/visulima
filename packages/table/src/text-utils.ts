import { stripVTControlCharacters } from "node:util";
import ansiRegex from "ansi-regex";
import stringWidth from "string-width";
import type { TruncateOptions } from "./types";

const globalAnsiPattern = ansiRegex();

/** Finds the real (nonâ€‘ANSI) character index in text corresponding to the visible position. */
function findRealPosition(text: string, visiblePosition: number): number {
    let visibleIndex = 0;
    let match: RegExpExecArray | null = null;
    const ansiRanges: { end: number; start: number }[] = [];
    globalAnsiPattern.lastIndex = 0;
    while ((match = globalAnsiPattern.exec(text)) !== null) {
        ansiRanges.push({ end: match.index + match[0].length, start: match.index });
    }
    let currentIndex = 0;
    while (currentIndex < text.length) {
        const range = ansiRanges.find((r) => currentIndex >= r.start && currentIndex < r.end);
        if (range) {
            currentIndex = range.end;
            continue;
        }
        const charWidth = stringWidth(text[currentIndex]);
        if (visibleIndex + charWidth > visiblePosition) {
            return currentIndex;
        }
        visibleIndex += charWidth;
        currentIndex++;
    }
    return Math.min(stringWidth(text), visiblePosition) - 1;
}

/** Finds the index of the nearest space character to a target position. */
function getIndexOfNearestSpace(text: string, targetIndex: number, searchRight = false): number {
    if (text.charAt(targetIndex) === " ") {
        return targetIndex;
    }
    const direction = searchRight ? 1 : -1;
    for (let offset = 0; offset <= 3; offset++) {
        const pos = targetIndex + offset * direction;
        if (text.charAt(pos) === " ") {
            return pos;
        }
    }
    return targetIndex;
}

/** Preserves ANSI codes in a text slice. */
function preserveAnsiCodes(text: string, startIndex: number, endIndex: number): string {
    const openCodes: string[] = [];
    let match: RegExpExecArray | null = null;
    globalAnsiPattern.lastIndex = 0;
    while ((match = globalAnsiPattern.exec(text)) !== null) {
        if (match.index > endIndex) {
            break;
        }
        const code = match[0];
        if (code === "\u001B[0m") {
            if (match.index < endIndex) {
                openCodes.length = 0;
            }
        } else if (code.startsWith("\u001B[") && match.index < endIndex) {
            openCodes.push(code);
        }
    }
    const slicedText = text.slice(startIndex, endIndex);
    return openCodes.join("") + slicedText + (openCodes.length > 0 ? "\u001B[0m" : "");
}

/** Truncates text to a specified width while preserving ANSI codes. */
export function truncateText(text: string, maxWidth: number, options: Required<TruncateOptions>): string {
    if (typeof text !== "string") {
        throw new TypeError(`Expected input to be a string, got ${typeof text}`);
    }
    if (typeof maxWidth !== "number") {
        throw new TypeError(`Expected maxWidth to be a number, got ${typeof maxWidth}`);
    }
    if (maxWidth < 1) {
        return "";
    }
    if (maxWidth === 1) {
        return options.truncationCharacter;
    }
    const visibleLength = stringWidth(text);
    if (visibleLength <= maxWidth) {
        return text;
    }
    const lines = text.split("\n");
    let { truncationCharacter } = options;
    const truncatedLines = lines.map((line) => {
        const lineLength = stringWidth(line);
        if (lineLength <= maxWidth) {
            return line;
        }
        if (options.position === "start") {
            if (options.preferTruncationOnSpace) {
                const nearestSpace = getIndexOfNearestSpace(line, lineLength - maxWidth + 1, true);
                return truncationCharacter + preserveAnsiCodes(line, nearestSpace, line.length).trim();
            }
            if (options.space) {
                truncationCharacter += " ";
            }
            const visibleStart = stringWidth(line) - maxWidth + stringWidth(truncationCharacter);
            const realStart = findRealPosition(line, visibleStart);
            return truncationCharacter + preserveAnsiCodes(line, realStart, line.length);
        }
        if (options.position === "middle") {
            if (options.space) {
                truncationCharacter = ` ${truncationCharacter} `;
            }
            const halfWidth = Math.floor(maxWidth / 2);
            if (options.preferTruncationOnSpace) {
                const spaceNearFirst = getIndexOfNearestSpace(line, halfWidth);
                const spaceNearSecond = getIndexOfNearestSpace(line, line.length - (maxWidth - halfWidth) + 1, true);
                return (
                    preserveAnsiCodes(line, 0, spaceNearFirst) +
                    truncationCharacter +
                    preserveAnsiCodes(line, spaceNearSecond, line.length).trim()
                );
            }
            const firstHalf = findRealPosition(line, halfWidth);
            const secondHalfStart = stringWidth(line) - (maxWidth - halfWidth) + stringWidth(truncationCharacter);
            const secondHalf = findRealPosition(line, secondHalfStart);
            return preserveAnsiCodes(line, 0, firstHalf) + truncationCharacter + preserveAnsiCodes(line, secondHalf, line.length);
        }
        if (options.position === "end") {
            if (options.preferTruncationOnSpace) {
                const nearestSpace = getIndexOfNearestSpace(line, maxWidth - 1);
                return preserveAnsiCodes(line, 0, nearestSpace) + truncationCharacter;
            }
            if (options.space) {
                truncationCharacter = ` ${truncationCharacter}`;
            }
            const realEnd = findRealPosition(line, maxWidth - stringWidth(truncationCharacter));
            return preserveAnsiCodes(line, 0, realEnd) + truncationCharacter;
        }
        throw new Error(`Expected options.position to be either 'start', 'middle' or 'end', got ${options.position}`);
    });
    return truncatedLines.join("\n");
}

/** Word wraps text to a specified width while preserving ANSI codes. */
export function wordWrapText(text: string, maxWidth: number): string[] {
    if (maxWidth <= 0 || !text) {
        return [text];
    }
    if (stringWidth(text) <= maxWidth) {
        return [text];
    }
    const lines = text.split(/\r?\n/);
    const wrappedLines: string[] = [];
    const colorPattern = ansiRegex();
    const linkPattern = /\u001B\]8;;([^\u0007]*)\u0007([^\u0007]*)\u001B\]8;;\u0007/g;
    for (const line of lines) {
        if (!line.trim()) {
            wrappedLines.push(line);
            continue;
        }
        const formats: { index: number; sequence: string }[] = [];
        let plainText = line;
        let match: RegExpExecArray | null = null;
        colorPattern.lastIndex = 0;
        while ((match = colorPattern.exec(line)) !== null) {
            formats.push({ index: match.index, sequence: match[0] });
        }
        linkPattern.lastIndex = 0;
        while ((match = linkPattern.exec(line)) !== null) {
            formats.push({
                index: match.index,
                sequence: `\u001B]8;;${match[1]}\u0007${match[2]}\u001B\]8;;\u0007`,
            });
        }
        formats.sort((a, b) => a.index - b.index);
        plainText = stripVTControlCharacters(line);
        const words = plainText.split(/\s+/);
        let currentLine = "";
        let currentLineWidth = 0;
        let lastAnsi = "";
        for (const word of words) {
            const wordWidth = stringWidth(word);
            if (currentLineWidth + wordWidth + (currentLine ? 1 : 0) > maxWidth && currentLine) {
                if (lastAnsi) {
                    currentLine += "\u001B[0m";
                }
                wrappedLines.push(currentLine);
                currentLine = "";
                currentLineWidth = 0;
            }
            if (currentLine) {
                currentLine += " ";
                currentLineWidth += 1;
            }
            if (lastAnsi) {
                currentLine += lastAnsi;
            }
            const wordStart = plainText.indexOf(word);
            let formattedWord = word;
            for (const format of formats) {
                if (format.index <= wordStart) {
                    if (format.sequence.startsWith("\u001B[")) {
                        lastAnsi = format.sequence;
                    }
                    formattedWord = format.sequence + formattedWord;
                }
            }
            currentLine += formattedWord;
            currentLineWidth += wordWidth;
        }
        if (currentLine) {
            if (lastAnsi) {
                currentLine += "\u001B[0m";
            }
            wrappedLines.push(currentLine);
        }
    }
    return wrappedLines;
}
