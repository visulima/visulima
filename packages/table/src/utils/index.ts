export { areCellsEquivalent } from "./are-cells-equivalent";
export { computeRowLogicalWidth } from "./compute-row-logical-width";
export { fillRowToWidth } from "./fill-row-to-width";
export { findRealPosition } from "./find-real-position";
export { getIndexOfNearestSpace } from "./get-index-of-nearest-space";
export { getRealCell } from "./get-real-cell";
export { preserveAnsiCodes } from "./preserve-ansi-codes";
export { truncateText } from "./truncate-text";
export { wordWrapText } from "./word-wrap-text";
