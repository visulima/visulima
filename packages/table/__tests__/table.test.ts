import { describe, expect, it } from "vitest";

import { Table } from "../src";
import { DEFAULT_BORDER, DOUBLE_BORDER, MARKDOWN_BORDER, MINIMAL_BORDER, ROUNDED_BORDER } from "../src/style";

describe("table", () => {
    describe("column width consistency", () => {
        it("should maintain consistent column widths across rows", () => {
            expect.assertions(7);

            const table = new Table();

            // Test case for the coverage report scenario
            table
                .setHeaders([
                    { content: "Component", hAlign: "center" },
                    { colSpan: 2, content: "Tests", hAlign: "center" },
                    { content: "Coverage", hAlign: "center" },
                ])
                .addRow([
                    { content: "Frontend", hAlign: "center", rowSpan: 2 },
                    { content: "Unit", hAlign: "left" },
                    { content: "156/156", hAlign: "right" },
                    { content: "100%", hAlign: "right" },
                ])
                .addRow([
                    null, // rowSpan from above
                    { content: "Integration", hAlign: "left" },
                    { content: "23/25", hAlign: "right" },
                    { content: "92%", hAlign: "right" },
                ]);

            const output = table.toString();
            const lines = output.split("\n");

            // Find the header line and data lines
            const headerLine = lines.find((line) => line.includes("Tests"))!;
            const unitLine = lines.find((line) => line.includes("Unit"))!;
            const integrationLine = lines.find((line) => line.includes("Integration"))!;

            // Extract cells from each line
            const unitCells = unitLine.split("│").map((cell) => cell.replaceAll(/\u001B\[\d+m/g, ""));
            const integrationCells = integrationLine.split("│").map((cell) => cell.replaceAll(/\u001B\[\d+m/g, ""));

            // Verify Unit and Integration cells have same width
            expect(unitCells[2].trim()).toBe("Unit");
            expect(integrationCells[2].trim()).toBe("Integration");
            expect(unitCells[2]).toHaveLength(integrationCells[2].length);

            // Verify coverage cells
            expect(unitCells[4].trim()).toBe("100%");
            expect(integrationCells[4].trim()).toBe("92%");

            // Additional verification for the entire table structure
            expect(lines).toEqual(
                expect.arrayContaining([
                    expect.stringContaining("┌───────────┬"),
                    expect.stringContaining("Component"),
                    expect.stringContaining("Tests"),
                    expect.stringContaining("Coverage"),
                    expect.stringContaining("Frontend"),
                    expect.stringContaining("Unit"),
                    expect.stringContaining("Integration"),
                    expect.stringContaining("└───────────┴"),
                ]),
            );

            // Verify the complete header structure
            const headerPattern = /^│\s*Component\s*│\s*Tests\s*│\s*Coverage\s*│/;
            expect(headerLine).toMatch(headerPattern);
        });

        it("should handle column spans while maintaining width consistency", () => {
            expect.assertions(1);

            const table = new Table();

            table
                .addRow([
                    { content: "Short", hAlign: "left" },
                    { colSpan: 2, content: "Very Long Content", hAlign: "center" },
                ])
                .addRow([
                    { content: "Test", hAlign: "left" },
                    { content: "Col1", hAlign: "left" },
                    { content: "Col2", hAlign: "left" },
                ]);

            const output = table.toString();
            const lines = output.split("\n");

            // Verify the table contains the expected content
            expect(lines).toEqual(
                expect.arrayContaining([
                    expect.stringContaining("Short"),
                    expect.stringContaining("Very Long Content"),
                    expect.stringContaining("Test"),
                    expect.stringContaining("Col1"),
                    expect.stringContaining("Col2"),
                ]),
            );
        });
    });

    describe("aNSI colored text", () => {
        it("should maintain correct padding with ANSI colored text", () => {
            expect.assertions(3);

            const table = new Table();
            const coloredText = "\u001B[32m✓ Passed\u001B[0m";
            const warningText = "\u001B[33m⚠ Warning\u001B[0m";

            table.setHeaders(["Status", "Value"]).addRow([coloredText, "100"]).addRow([warningText, "50"]);

            const output = table.toString();
            const lines = output.split("\n");

            // Verify structure
            expect(lines).toEqual(
                expect.arrayContaining([
                    expect.stringContaining("Status"),
                    expect.stringContaining("Value"),
                    expect.stringContaining("✓"),
                    expect.stringContaining("⚠"),
                ]),
            );

            // Verify colored text is preserved
            expect(output).toContain("\u001B[32m");
            expect(output).toContain("\u001B[33m");
        });
    });

    describe("unicode and emoji support", () => {
        it("should handle Unicode characters and emojis correctly", () => {
            expect.assertions(1);

            const table = new Table();

            table.setHeaders(["📊 Metrics", "Value"]).addRow(["代码覆盖率", "80%"]).addRow(["性能测试", "95%"]);

            const output = table.toString();
            const lines = output.split("\n");

            // Verify content
            expect(lines).toEqual(
                expect.arrayContaining([
                    expect.stringContaining("📊"),
                    expect.stringContaining("代码覆盖率"),
                    expect.stringContaining("性能测试"),
                    expect.stringContaining("80%"),
                    expect.stringContaining("95%"),
                ]),
            );
        });
    });

    describe("truncation and multi-line text", () => {
        it("should truncate long text and handle multi-line content", () => {
            expect.assertions(3);

            const table = new Table({ maxWidth: 20, truncate: true });
            const longText = "This is a very long text that should be truncated";
            const multiLineText = "First line\nSecond line\nThird line";

            table.addRow([{ content: longText }]).addRow([{ content: multiLineText }]);

            const output = table.toString();
            const lines = output.split("\n");

            // Verify truncation
            const truncatedLine = lines.find((line) => line.includes("This"));

            expect(truncatedLine).toBeDefined();
            expect((truncatedLine as string).length).toBeLessThanOrEqual(24); // 20 + padding + borders

            // Verify multi-line
            expect(lines).toEqual(
                expect.arrayContaining([expect.stringContaining("First line"), expect.stringContaining("Second line"), expect.stringContaining("Third line")]),
            );
        });

        it("should properly truncate colored text", () => {
            expect.assertions(6);

            const table = new Table({ maxWidth: 20, truncate: true });
            const coloredText = "\u001B[32mThis is a very long green text that should be truncated\u001B[0m";

            table.addRow([{ content: coloredText }]);

            const output = table.toString();
            const lines = output.split("\n");

            // Find the line containing the colored text (including ANSI codes)
            const truncatedLine = lines.find((line) => line.includes("\u001B[32m"));
            expect(truncatedLine).toBeDefined();

            // Verify the line contains both start and end color codes
            expect(truncatedLine).toContain("\u001B[32m");
            expect(truncatedLine).toContain("\u001B[0m");

            // Verify length (excluding color codes)
            const strippedLine = (truncatedLine as string).replaceAll(/\u001B\[\d+m/g, "");
            expect(strippedLine.trim().length).toBeLessThanOrEqual(24); // 20 + padding + borders

            // Verify that the ellipsis is colored
            const lastColorStart = truncatedLine?.lastIndexOf("\u001B[32m");
            const lastColorEnd = truncatedLine?.lastIndexOf("\u001B[0m");
            expect(lastColorStart).toBeLessThan(lastColorEnd);
            expect(truncatedLine?.slice(lastColorEnd! - 3, lastColorEnd)).toBe("...");
        });
    });

    describe("border styles", () => {
        it.each([
            ["default", DEFAULT_BORDER],
            ["minimal", MINIMAL_BORDER],
            ["markdown", MARKDOWN_BORDER],
            ["double", DOUBLE_BORDER],
            ["rounded", ROUNDED_BORDER],
        ])("should support %s border style", (styleName, borderStyle) => {
            expect.assertions(1);

            const table = new Table({ border: borderStyle });
            table.setHeaders(["Header 1", "Header 2"]);
            table.addRow(["Value 1", "Value 2"]);

            const output = table.toString();

            expect(output).toMatchSnapshot(`table-${styleName}-style`);
        });
    });

    describe("cell alignment", () => {
        it("should handle different cell alignments", () => {
            expect.assertions(4);

            const table = new Table({ padding: 1 });
            const content = "test";

            table.addRow([
                { content, hAlign: "left" },
                { content, hAlign: "center" },
                { content, hAlign: "right" },
            ]);

            const output = table.toString();
            const lines = output.split("\n");
            const dataLine = lines.find((line) => line.includes("test"))!;

            // Verify content is present
            expect(output).toContain("test");

            // Verify each alignment
            const cells = dataLine.split("│").slice(1, -1);
            expect(cells[0].trim()).toBe("test"); // Left aligned
            expect(cells[1].trim()).toBe("test"); // Center aligned
            expect(cells[2].trim()).toBe("test"); // Right aligned
        });
    });

    describe("padding configuration", () => {
        it("should respect custom padding settings", () => {
            expect.assertions(2);

            const table = new Table({ padding: 3 });
            table.addRow(["Test"]);

            const output = table.toString();
            const lines = output.split("\n");
            const dataLine = lines.find((line) => line.includes("Test"))!;

            // Verify content is present with padding
            const cell = dataLine.split("│")[1];
            expect(cell.startsWith("   ")).toBeTruthy();
            expect(cell.endsWith("   ")).toBeTruthy();
        });
    });

    describe("empty and null values", () => {
        it("should handle empty table with no headers", () => {
            expect.assertions(1);

            const table = new Table();
            table.addRow(["", { content: "" }, { content: null }]).addRow([" ", { content: undefined }, { content: "" }]);

            const output = table.toString();

            expect(output).toMatchInlineSnapshot(`
                "┌───┬───┬───┐
                │   │   │   │
                ├───┼───┼───┤
                │   │   │   │
                └───┴───┴───┘
                "
            `);
        });

        it("should handle table with empty headers", () => {
            expect.assertions(1);

            const table = new Table();
            table
                .setHeaders(["", null, undefined])
                .addRow(["", { content: "" }, { content: null }])
                .addRow([" ", { content: undefined }, { content: "" }]);

            const output = table.toString();

            expect(output).toMatchInlineSnapshot(`
                "┌───┬───┬───┐
                │   │   │   │
                ├───┼───┼───┤
                │   │   │   │
                ├───┼───┼───┤
                │   │   │   │
                └───┴───┴───┘
                "
            `);
        });

        it("should handle table with hidden headers", () => {
            expect.assertions(1);

            const table = new Table({ showHeader: false });
            table
                .setHeaders(["", null, undefined])
                .addRow(["", { content: "" }, { content: null }])
                .addRow([" ", { content: undefined }, { content: "" }]);

            const output = table.toString();

            expect(output).toMatchInlineSnapshot(`
                "┌───┬───┬───┐
                │   │   │   │
                ├───┼───┼───┤
                │   │   │   │
                └───┴───┴───┘
                "
            `);
        });
    });

    describe("table creation and options", () => {
        it("should handle various constructor options", () => {
            expect.assertions(3);

            const table = new Table({
                align: "center",
                border: ROUNDED_BORDER,
                maxWidth: 50,
                padding: 2,
                showHeader: false,
                truncate: true,
            });

            table.addRow(["Test"]);
            const output = table.toString();

            expect(output).toContain("╭"); // Rounded border
            expect(output).toContain("  Test  "); // Padding of 2
            expect(output).not.toContain("undefined"); // No header
        });

        it("should validate constructor options", () => {
            expect.assertions(2);

            expect(() => new Table({ padding: -1 })).toThrow("padding must be a non-negative number");
            expect(() => new Table({ maxWidth: 0 })).toThrow("maxWidth must be a positive number");
        });
    });

    describe("complex tables", () => {
        it("should handle complex table with progress bars, colors, and Unicode", () => {
            expect.assertions(6);

            const table = new Table();
            const blue = (text: string) => `\u001B[34m${text}\u001B[0m`;
            const green = (text: string) => `\u001B[32m${text}\u001B[0m`;
            const red = (text: string) => `\u001B[31m${text}\u001B[0m`;

            table
                .setHeaders([
                    { content: "📊 Metrics", hAlign: "center" },
                    { content: "📈 Progress", hAlign: "center" },
                    { content: "🎯 Target", hAlign: "center" },
                ])
                .addRow([
                    { content: "代码覆盖率", hAlign: "left" },
                    { content: blue("▓▓▓▓▓▓▓▓░░ 80%"), hAlign: "center" },
                    { content: "90%", hAlign: "right" },
                ])
                .addRow([
                    { content: "性能测试", hAlign: "left" },
                    { content: green("▓▓▓▓▓▓▓▓▓░ 95%"), hAlign: "center" },
                    { content: "85%", hAlign: "right" },
                ])
                .addRow([
                    { content: "安全扫描", hAlign: "left" },
                    { content: red("▓▓▓▓░░░░░░ 40%"), hAlign: "center" },
                    { content: "100%", hAlign: "right" },
                ]);

            const output = table.toString();

            // Log the table for visual inspection during test development
            console.log("\nComplex Table Output:\n" + output + "\n");

            const lines = output.split("\n");

            // Verify header alignment
            const headerLine = lines.find((line) => line.includes("📊 Metrics"));
            expect(headerLine).toBeDefined();
            expect(headerLine).toContain("📊 Metrics"); // Center aligned header

            // Verify progress bar colors
            const blueLine = lines.find((line) => line.includes("80%"));
            const greenLine = lines.find((line) => line.includes("95%"));
            const redLine = lines.find((line) => line.includes("40%"));

            expect(blueLine).toContain("\u001B[34m"); // Blue color
            expect(greenLine).toContain("\u001B[32m"); // Green color
            expect(redLine).toContain("\u001B[31m"); // Red color

            // Verify overall table layout
            expect(output).toMatchSnapshot("complex-table-with-progress");
        });
    });
});
