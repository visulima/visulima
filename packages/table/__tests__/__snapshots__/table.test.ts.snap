// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`table > border styles > should support default border style > table-default-style 1`] = `
"┌──────────┬──────────┐
│ Header 1 │ Header 2 │
├──────────┼──────────┤
│ Value 1  │ Value 2  │
└──────────┴──────────┘
"
`;

exports[`table > border styles > should support double border style > table-double-style 1`] = `
"╔══════════╦══════════╗
║ Header 1 ║ Header 2 ║
╠══════════╬══════════╣
║ Value 1  ║ Value 2  ║
╚══════════╩══════════╝
"
`;

exports[`table > border styles > should support markdown border style > table-markdown-style 1`] = `
"|----------|----------|
| Header 1 | Header 2 |
|----------|----------|
| Value 1  | Value 2  |
|----------|----------|
"
`;

exports[`table > border styles > should support minimal border style > table-minimal-style 1`] = `
"┌──────────┬──────────┐
│ Header 1 │ Header 2 │
│ Value 1  │ Value 2  │
└──────────┴──────────┘
"
`;

exports[`table > border styles > should support rounded border style > table-rounded-style 1`] = `
"╭──────────┬──────────╮
│ Header 1 │ Header 2 │
├──────────┼──────────┤
│ Value 1  │ Value 2  │
╰──────────┴──────────╯
"
`;

exports[`table > complex tables > should handle complex table with progress bars, colors, and Unicode > complex-table-with-progress 1`] = `
"┌────────────┬────────────────┬───────────┐
│ 📊 Metrics │  📈 Progress   │ 🎯 Target │
├────────────┼────────────────┼───────────┤
│ 代码覆盖率 │ [34m▓▓▓▓▓▓▓▓░░ 80%[0m │       90% │
├────────────┼────────────────┼───────────┤
│ 性能测试   │ [32m▓▓▓▓▓▓▓▓▓░ 95%[0m │       85% │
├────────────┼────────────────┼───────────┤
│ 安全扫描   │ [31m▓▓▓▓░░░░░░ 40%[0m │      100% │
└────────────┴────────────────┴───────────┘
"
`;
